### PR TITLE
Kafka

### LINKED ISSUE #18: Support kafka initialization

### PR DESCRIPTION
**Pull request by alkiskal**
Fixes #18 
Adds kafka support, example: https://github.com/Workable/maestro/pull/266

Next step: Make kafka dependencies optional through codependence.

### CODE REVIEW

File: examples/kafka-example/routes.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  brokers: config.kafka.brokers
+};
+const kafka = new Kafka(options);
+
+module.exports = {
+  get: {
+    '/read': async (ctx, next) => {
+      await kafka.connect();
+      const topic = config.kafka.consumer.topics.test;
+      new KafkaHandler(kafka, { topic });
```
Comment (nikostoulas) [inline examples/kafka-example/routes.js]: 👍

File: src/initializers/kafka/base-kafka-handler.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  constructor(kafka: Kafka, options: { topic: string; logger: Logger; batchSize?: number; }) {
+    const { topic, batchSize, logger } = options;
+    this.topic = topic;
+    this.batchSize = batchSize || 1;
+    this.logger = logger;
+    this.consumer = kafka.createConsumer(this.topic);
+    this.consumer.connect().then(() => this.consume().catch(err => this.logger.error(err)));
+  }
+
+  async abstract handle(message: any): Promise<any>;
```
Comment (nikostoulas) [inline src/initializers/kafka/base-kafka-handler.ts]: Can you add typings in message? Or is it only the value?

Comment (alkiskal) [reply src/initializers/kafka/base-kafka-handler.ts]: I'll have to check. For now I need only the ok for the interface

Comment (nikostoulas) [reply src/initializers/kafka/base-kafka-handler.ts]: The example is clear. I like it

Comment (alkiskal) [reply src/initializers/kafka/base-kafka-handler.ts]: This is only the value, `await Promise.all(flatten([messages]).map(m => this.handle(m.value)));`
Moreover, this is already an object that we can use 
```
      false,
      true, //Receive as object
```

Comment (alkiskal) [reply src/initializers/kafka/base-kafka-handler.ts]: I believe we can close this for now. Maestro has a working example, we can come back for any improvements if the need arises.

File: src/initializers/kafka/base-kafka-handler.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  batchSize: number;
+  logger: Logger;
+
+  constructor(kafka: Kafka, options: { topic: string; logger: Logger; batchSize?: number; }) {
+    const { topic, batchSize, logger } = options;
+    this.topic = topic;
+    this.batchSize = batchSize || 1;
+    this.logger = logger;
+    this.consumer = kafka.createConsumer(this.topic);
+    this.consumer.connect().then(() => this.consume().catch(err => this.logger.error(err)));
```
Comment (anastasios-workable) [inline src/initializers/kafka/base-kafka-handler.ts]: If this fails the instance is invalid. Can you move object instantiation to an async factory method instead?

File: src/initializers/kafka/kafka.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  public createConsumer(topic: string) {
+    const { groupId, brokers } = this.options;
+    const config = {
+      groupId,
+      logger: getLogger('orka.kafka.consumer.internal'),
+      noptions: {
+        'group.id': groupId,
+        'metadata.broker.list': brokers.join(','),
+        'socket.keepalive.enable': true,
```
Comment (anastasios-workable) [inline src/initializers/kafka/kafka.ts]: `Socket keepalives need to be enabled on both sides to expedite connection termination on connectivity loss (e.g., network partitioning) for both parties. This means the broker needs to implement it too.` Does it?

File: src/initializers/kafka/auth-options.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  };
+  logger.trace('Certificates created: ', paths);
+  return paths;
+};
+
+export default ({ key, cert, ca }) => {
+  const paths = certificates({ key, cert, ca });
+
+  return {
+    'security.protocol': <'ssl'> 'ssl',
```
[the remark this answers is not in the document]
Comment (alkiskal) [reply src/initializers/kafka/auth-options.ts]: It's a workaround for enums, so you can use string.

File: src/initializers/kafka/base-kafka-handler.ts (hunk 3 above)
Comment (alkiskal) [reply src/initializers/kafka/base-kafka-handler.ts]: I can remove the `consumer.connect()` from the constructor and add it in the `consume`. Then the caller must call `consume()` explicitly. Does that work?

File: src/initializers/kafka/kafka.ts (hunk 4 above)
Comment (alkiskal) [reply src/initializers/kafka/kafka.ts]: I can't find anything related in lenses. I found in the issues section that it doesn't cause any harm if it's enabled. We can check further if you want, but I'll need assistance regarding the lenses/kafka configuration.

**Comment by alkiskal** [comment]
@klesgidis It does not. https://github.com/nodefluent/node-sinek#you-might-also-like . But we don't need it. In librdkafka we used streams because it was easier to use and it handled reconnects etc, but in this it handles all that already. Plus the native consumer/producer are faster.

File: src/initializers/kafka/kafka.ts (hunk 4 above)
Comment (anastasios-workable) [reply src/initializers/kafka/kafka.ts]: Thanks for checking.

File: src/initializers/kafka/base-kafka-handler.ts (hunk 3 above)
Comment (nikostoulas) [reply src/initializers/kafka/base-kafka-handler.ts]: What if the connection fails after a while? Can we handle reconnections somehow?

Comment (alkiskal) [reply src/initializers/kafka/base-kafka-handler.ts]: Supposedly it handles reconnections automatically
```
      this.consumer.on("disconnected", () => {
        if(this._inClosing){
          this._reset();
        }
        logger.warn("Disconnected.");
        //auto-reconnect --> handled by consumer.consume();
      });
```

I'll have to test locally (terminate the broker and bring it back up or something equivalent)

Comment (alkiskal) [reply src/initializers/kafka/base-kafka-handler.ts]: Used this https://hub.docker.com/r/landoop/kafka-lenses-dev/ locally in order to test what happens when terminating the broker. 
Everything works as expected. Both the consuming and producing resume automatically.

**Comment by nikostoulas** [comment]
Can we start adding some unit tests?

**Pull request merged**